File: app/src/main/java/me/ykrank/s1next/view/dialog/RateRequestDialogFragment.java
package me.ykrank.s1next.view.dialog;

import android.os.Bundle;
import android.support.annotation.NonNull;

import io.reactivex.Observable;
import me.ykrank.s1next.App;
import me.ykrank.s1next.R;
import me.ykrank.s1next.data.api.model.RatePreInfo;
import me.ykrank.s1next.data.api.model.RateResult;
import me.ykrank.s1next.widget.track.event.NewRateTrackEvent;
import me.ykrank.s1next.widget.track.event.page.PageEndEvent;
import me.ykrank.s1next.widget.track.event.page.PageStartEvent;

/**
 * A dialog requests to reply to post.
 */
public final class RateRequestDialogFragment extends ProgressDialogFragment<RateResult> {

    public static final String TAG = RateRequestDialogFragment.class.getName();

    private static final String ARG_RATE_PRE_INFO = "rate_pre_info";
    private static final String ARG_SCORE = "score";
    private static final String ARG_REASON = "reason";

    private static final String STATUS_REPLY_SUCCESS = "post_reply_succeed";

    public static RateRequestDialogFragment newInstance(@NonNull RatePreInfo ratePreInfo, String score, String reason) {
        App.get().getTrackAgent().post(new NewRateTrackEvent(ratePreInfo.getTid(), ratePreInfo.getPid(), score, reason));

        RateRequestDialogFragment fragment = new RateRequestDialogFragment();
        Bundle bundle = new Bundle();
        bundle.putParcelable(ARG_RATE_PRE_INFO, ratePreInfo);
        bundle.putString(ARG_SCORE, score);
        bundle.putString(ARG_REASON, reason);
        fragment.setArguments(bundle);

        return fragment;
    }

    @Override
    protected CharSequence getProgressMessage() {
        return getText(R.string.dialog_progress_message_reply);
    }

    @Override
    protected Observable<RateResult> getSourceObservable() {
        RatePreInfo ratePreInfo = getArguments().getParcelable(ARG_RATE_PRE_INFO);
        String score = getArguments().getString(ARG_SCORE);
        String reason = getArguments().getString(ARG_REASON);
        if (ratePreInfo == null) {
            return Observable.error(new IllegalStateException("RatePreInfo is null"));
        }
        return mS1Service.rate(ratePreInfo.getFormHash(), ratePreInfo.getTid(), ratePreInfo.getPid(),
                ratePreInfo.getRefer(), ratePreInfo.getHandleKey(), score, reason)
                .map(RateResult::fromHtml);
    }

    @Override
    protected void onNext(RateResult data) {
        if (data.isSuccess()) {
            showShortTextAndFinishCurrentActivity(getString(R.string.rate_success));
        } else {
            showShortText(data.getErrorMsg());
        }
    }

    @Override
    public void onResume() {
        super.onResume();
        trackAgent.post(new PageStartEvent(getContext(), "弹窗-回复进度条-"));
    }

    @Override
    public void onPause() {
        trackAgent.post(new PageEndEvent(getContext(), "弹窗-回复进度条-"));
        super.onPause();
    }
}

File: app/src/main/java/me/ykrank/s1next/view/dialog/requestdialog/RateRequestDialogFragment.kt
package me.ykrank.s1next.view.dialog.requestdialog

import android.os.Bundle
import io.reactivex.Single
import me.ykrank.s1next.App
import me.ykrank.s1next.R
import me.ykrank.s1next.data.api.model.AjaxResult
import me.ykrank.s1next.data.api.model.RatePreInfo
import me.ykrank.s1next.view.dialog.ProgressDialogFragment
import me.ykrank.s1next.widget.track.event.NewRateTrackEvent

/**
 * A dialog requests to reply to post.
 */
class RateRequestDialogFragment : ProgressDialogFragment<AjaxResult>() {

    override fun getProgressMessage(): CharSequence? {
        return getText(R.string.dialog_progress_message_reply)
    }

    override fun getSourceObservable(): Single<AjaxResult> {
        val bundle = requireArguments()
        val ratePreInfo = bundle.getParcelable<RatePreInfo>(ARG_RATE_PRE_INFO)
        val score = bundle.getString(ARG_SCORE)
        val reason = bundle.getString(ARG_REASON)
        if (ratePreInfo == null) {
            return Single.error<AjaxResult>(IllegalStateException("RatePreInfo is null"))
        }
        return mS1Service.rate(ratePreInfo.formHash, ratePreInfo.tid, ratePreInfo.pid,
                ratePreInfo.refer, ratePreInfo.handleKey, score, reason)
            .map { AjaxResult.fromAjaxString(it) }
    }

    override fun onNext(data: AjaxResult) {
        if (data.success) {
            showShortTextAndFinishCurrentActivity(getString(R.string.rate_success))
        } else {
            showToastText(data.msg)
        }
    }

    companion object {

        val TAG: String = RateRequestDialogFragment::class.java.name

        private const val ARG_RATE_PRE_INFO = "rate_pre_info"
        private const val ARG_SCORE = "score"
        private const val ARG_REASON = "reason"

        fun newInstance(ratePreInfo: RatePreInfo, score: String, reason: String): RateRequestDialogFragment {
            App.get().trackAgent.post(NewRateTrackEvent(ratePreInfo.tid, ratePreInfo.pid, score, reason))

            val fragment = RateRequestDialogFragment()
            val bundle = Bundle()
            bundle.putParcelable(ARG_RATE_PRE_INFO, ratePreInfo)
            bundle.putString(ARG_SCORE, score)
            bundle.putString(ARG_REASON, reason)
            fragment.arguments = bundle

            return fragment
        }
    }
}
